File: src/engine/downloader.ts
import {spawn, type ChildProcess} from 'node:child_process'
import fs from 'node:fs/promises'
import path from 'node:path'
import {hasMutagen} from './binary.js'
import {parseAriaProgressLine} from './aria2c.js'
import type {Portion} from './extractor.js'

export type BakeProgress = {
  downloadedBytes: number
  totalBytes?: number
  speed?: number
  eta?: number
  part: number
  totalParts: number
  playlistItem?: number
  playlistTotal?: number
  connections?: number
  seeders?: number
}

export type BakeHandlers = {
  onProgress: (progress: BakeProgress) => void
  onProcessing: () => void
}

const PROGRESS_TAG = 'ANPAN|'
const PROGRESS_TEMPLATE = `${PROGRESS_TAG}%(progress.downloaded_bytes)s|%(progress.total_bytes)s|%(progress.total_bytes_estimate)s|%(progress.speed)s|%(progress.eta)s`

let activeChild: ChildProcess | undefined
process.on('exit', () => activeChild?.kill('SIGTERM'))

export type BakeVideoOptions = {
  ytdlpBin: string
  ffmpegLocation?: string
  aria2cArgs?: string[]
  url: string
  cachedJsonPath?: string
  portion: Portion
  outputDir: string
  timeRange?: string
  isPlaylist?: boolean
  cookiesBrowser?: string
  subtitles?: 'off' | 'embed' | 'write'
  subLangs?: string
  sponsorBlock?: 'off' | 'remove' | 'mark'
  writeThumbnail?: boolean
}

export async function bakeVideo(
  opts: BakeVideoOptions,
  handlers: BakeHandlers,
  signal?: AbortSignal,
): Promise<string> {
  const outputTemplate = opts.isPlaylist
    ? path.join(
        opts.outputDir,
        '%(playlist_title,playlist)s',
        '%(playlist_index)02d - %(title)s.%(ext)s',
      )
    : path.join(
        opts.outputDir,
        `%(title)s${opts.timeRange ? ` [${opts.timeRange.replace(/[:*]/g, '.')}]` : ''}.%(ext)s`,
      )

  // Opus and Ogg require python-mutagen for embedding thumbnails in yt-dlp.
  // If mutagen is not installed, yt-dlp hard-crashes. We gracefully drop --embed-thumbnail
  // so the download succeeds with full audio quality and metadata.
  const isOpusOrOgg =
    opts.portion.label.toLowerCase().includes('opus') ||
    opts.portion.label.toLowerCase().includes('ogg')
  let portionArgs = opts.portion.ytdlpArgs
  if (isOpusOrOgg && !(await hasMutagen(opts.ytdlpBin))) {
    portionArgs = portionArgs.filter(arg => arg !== '--embed-thumbnail')
  }

  const args = [
    ...(opts.cachedJsonPath && !opts.isPlaylist ? ['--load-info-json', opts.cachedJsonPath] : [opts.url]),
    ...portionArgs,
    ...(opts.isPlaylist
      ? [
          '--yes-playlist',
          '--ignore-errors',
          '--replace-in-metadata',
          'playlist_title',
          '(?i)^(?:album|ep|single)\\s*[-:–—]\\s*',
          '',
          '--replace-in-metadata',
          'playlist',
          '(?i)^(?:album|ep|single)\\s*[-:–—]\\s*',
          '',
        ]
      : ['--no-playlist']),
    '--no-warnings',
    '--newline',
    '--no-quiet',
    '--progress',
    '--progress-template',
    `download:${PROGRESS_TEMPLATE}`,
    // --print after_move:filepath writes the resolved final file destination
    // after container merging and postprocessing completes.
    '--print',
    'after_move:filepath',
    '--no-simulate',
    '-o',
    outputTemplate,
  ]
  if (opts.timeRange) {
    args.push('--download-sections', `*${opts.timeRange}`)
  }
  if (opts.ffmpegLocation) args.push('--ffmpeg-location', opts.ffmpegLocation)
  if (opts.aria2cArgs) args.push(...opts.aria2cArgs)

  // Configurable yt-dlp parameters
  if (opts.cookiesBrowser && opts.cookiesBrowser !== 'none') {
    args.push('--cookies-from-browser', opts.cookiesBrowser)
  }
  if (opts.sponsorBlock === 'remove') {
    args.push('--sponsorblock-remove', 'all')
  } else if (opts.sponsorBlock === 'mark') {
    args.push('--sponsorblock-mark', 'all')
  }
  if (opts.subtitles === 'embed') {
    args.push('--embed-subs', '--sub-langs', opts.subLangs || 'vi,en')
  } else if (opts.subtitles === 'write') {
    args.push('--write-subs', '--sub-langs', opts.subLangs || 'vi,en')
  }
  if (opts.writeThumbnail) {
    args.push('--write-thumbnail')
  }

  await fs.mkdir(opts.outputDir, {recursive: true})

  return new Promise((resolve, reject) => {
    const child = spawn(opts.ytdlpBin, args, {signal})
    activeChild = child

    let stderr = ''
    let filepath = ''
    let part = 0
    let totalParts = 1
    let lastDownloaded = 0
    let buffer = ''
    let playlistItem: number | undefined
    let playlistTotal: number | undefined
    const destinations: string[] = []

    child.stdout.on('data', (chunk: Buffer) => {
      buffer += chunk.toString()
      const lines = buffer.split('\n')
      buffer = lines.pop() ?? ''

      for (const rawLine of lines) {
        const line = rawLine.trim()
        if (!line) continue

        const itemMatch = /^\[download\] Downloading item (\d+) of (\d+)/.exec(line)
        if (itemMatch) {
          playlistItem = Number.parseInt(itemMatch[1]!, 10)
          playlistTotal = Number.parseInt(itemMatch[2]!, 10)
        }

        // Native yt-dlp progress template
        if (line.startsWith(PROGRESS_TAG)) {
          const [downloaded, total, totalEstimate, speed, eta] = line
            .slice(PROGRESS_TAG.length)
            .split('|')
          const downloadedBytes = parseNumber(downloaded) ?? 0
          if (downloadedBytes < lastDownloaded) part++
          lastDownloaded = downloadedBytes
          handlers.onProgress({
            downloadedBytes,
            totalBytes: parseNumber(total) ?? parseNumber(totalEstimate),
            speed: parseNumber(speed),
            eta: parseNumber(eta),
            part,
            totalParts,
            playlistItem,
            playlistTotal,
          })
        } else {
          const aria = parseAriaProgressLine(line)
          if (aria) {
            if (aria.downloadedBytes < lastDownloaded && aria.downloadedBytes > 0) part++
            lastDownloaded = aria.downloadedBytes
            handlers.onProgress({
              downloadedBytes: aria.downloadedBytes,
              totalBytes: aria.totalBytes,
              speed: aria.speed,
              eta: aria.eta,
              part,
              totalParts,
              playlistItem,
              playlistTotal,
              connections: aria.connections,
              seeders: aria.seeders,
            })
          }
        }

        if (line.includes('Downloading 1 format(s):')) {
          totalParts = (line.split('format(s):')[1] ?? '').trim().split('+').length
        } else if (line.includes('[Merger]') || line.includes('[ExtractAudio]')) {
          const merging = /^\[Merger\] Merging formats into "(.+)"$/.exec(line)?.[1]
          const extracting = /^\[ExtractAudio\] Destination: (.+)$/.exec(line)?.[1]
          const target = merging ?? extracting
          if (target) destinations.push(target)
          handlers.onProcessing()
        } else if (line.startsWith('[download] Destination: ')) {
          destinations.push(line.slice('[download] Destination: '.length))
        } else if (path.isAbsolute(line)) {
          filepath = line
        }
      }
    })

    child.stderr.on('data', (chunk: Buffer) => (stderr += chunk))
    child.on('error', reject)
    child.on('close', code => {
      activeChild = undefined
      if (signal?.aborted) {
        void removePartials(destinations)
        reject(new Error('Download cancelled.'))
        return
      }
      if (code === 0 && filepath) {
        resolve(opts.isPlaylist ? path.dirname(filepath) : filepath)
      } else {
        reject(new Error(cleanErrorOutput(stderr) || `Download failed (yt-dlp exit code ${code}).`))
      }
    })
  })
}

function removePartials(destinations: string[]): Promise<unknown> {
  return Promise.allSettled(
    destinations
      .flatMap(dest => [dest, `${dest}.part`, `${dest}.ytdl`])
      .map(file => fs.rm(file, {force: true})),
  )
}

function parseNumber(value: string | undefined): number | undefined {
  if (!value || value === 'NA' || value === 'None') return undefined
  const n = Number.parseFloat(value)
  return Number.isFinite(n) ? n : undefined
}

function cleanErrorOutput(stderr: string): string {
  const lines = stderr
    .split('\n')
    .map(l => l.trim())
    .filter(l => l.startsWith('ERROR:'))
  const last = lines.at(-1)
  return last ? last.replace(/^ERROR:\s*(\[[^\]]+\]\s*)?/, '') : ''
}

File: internal/engine/downloader.go
package engine

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

const progressTag = "ANPAN|"
const progressTemplate = progressTag + "%(progress.downloaded_bytes)s|%(progress.total_bytes)s|%(progress.total_bytes_estimate)s|%(progress.speed)s|%(progress.eta)s"

type BakeVideoOptions struct {
	YtdlpBin       string
	FfmpegLocation string
	Aria2cArgs     []string
	URL            string
	CachedJSONPath string
	Portion        Portion
	OutputDir      string
	TimeRange      string
	IsPlaylist     bool
	CookiesBrowser string
	Subtitles      string // "off" | "embed" | "write"
	SubLangs       string
	SponsorBlock   string // "off" | "remove" | "mark"
	WriteThumbnail bool
}

func parseProgressFloat(v string) *float64 {
	clean := strings.TrimSpace(v)
	if clean == "" || clean == "NA" || clean == "None" {
		return nil
	}
	if f, err := strconv.ParseFloat(clean, 64); err == nil {
		return &f
	}
	return nil
}

var (
	playlistItemRegex    = regexp.MustCompile(`^\[download\] Downloading item (\d+) of (\d+)`)
	formatCountRegex     = regexp.MustCompile(`Downloading 1 format\(s\):\s*(.+)`)
	mergerTargetRegex    = regexp.MustCompile(`^\[Merger\] Merging formats into "(.+)"$`)
	extractAudioTargetRegex = regexp.MustCompile(`^\[ExtractAudio\] Destination: (.+)$`)
)

func BakeVideo(ctx context.Context, opts BakeVideoOptions, handlers BakeHandlers) (string, error) {
	var outputTemplate string
	if opts.IsPlaylist {
		outputTemplate = filepath.Join(opts.OutputDir, "%(playlist_title,playlist)s", "%(playlist_index)02d - %(title)s.%(ext)s")
	} else {
		trimSuffix := ""
		if opts.TimeRange != "" {
			cleanRange := strings.ReplaceAll(strings.ReplaceAll(opts.TimeRange, ":", "."), "*", ".")
			trimSuffix = fmt.Sprintf(" [%s]", cleanRange)
		}
		outputTemplate = filepath.Join(opts.OutputDir, fmt.Sprintf("%%(title)s%s.%%(ext)s", trimSuffix))
	}

	portionArgs := opts.Portion.YtdlpArgs
	isOpusOrOgg := strings.Contains(strings.ToLower(opts.Portion.Label), "opus") || strings.Contains(strings.ToLower(opts.Portion.Label), "ogg")
	if isOpusOrOgg && !HasMutagen(opts.YtdlpBin) {
		var filtered []string
		for _, a := range portionArgs {
			if a != "--embed-thumbnail" {
				filtered = append(filtered, a)
			}
		}
		portionArgs = filtered
	}

	var args []string
	if opts.CachedJSONPath != "" && !opts.IsPlaylist {
		args = append(args, "--load-info-json", opts.CachedJSONPath)
	} else {
		args = append(args, opts.URL)
	}
	args = append(args, portionArgs...)

	if opts.IsPlaylist {
		args = append(args,
			"--yes-playlist",
			"--ignore-errors",
			"--replace-in-metadata", "playlist_title", "(?i)^(?:album|ep|single)\\s*[-:–—]\\s*", "",
			"--replace-in-metadata", "playlist", "(?i)^(?:album|ep|single)\\s*[-:–—]\\s*", "",
		)
	} else {
		args = append(args, "--no-playlist")
	}

	args = append(args,
		"--no-warnings",
		"--newline",
		"--no-quiet",
		"--progress",
		"--progress-template", "download:"+progressTemplate,
		"--print", "after_move:filepath",
		"--no-simulate",
		"-o", outputTemplate,
	)

	if opts.TimeRange != "" {
		args = append(args, "--download-sections", "*"+opts.TimeRange)
	}
	if opts.FfmpegLocation != "" {
		args = append(args, "--ffmpeg-location", opts.FfmpegLocation)
	}
	if len(opts.Aria2cArgs) > 0 {
		args = append(args, opts.Aria2cArgs...)
	}
	if opts.CookiesBrowser != "" && opts.CookiesBrowser != "none" {
		args = append(args, "--cookies-from-browser", opts.CookiesBrowser)
	}
	if opts.SponsorBlock == "remove" {
		args = append(args, "--sponsorblock-remove", "all")
	} else if opts.SponsorBlock == "mark" {
		args = append(args, "--sponsorblock-mark", "all")
	}
	if opts.Subtitles == "embed" {
		subLangs := "vi,en"
		if opts.SubLangs != "" {
			subLangs = opts.SubLangs
		}
		args = append(args, "--embed-subs", "--sub-langs", subLangs)
	} else if opts.Subtitles == "write" {
		subLangs := "vi,en"
		if opts.SubLangs != "" {
			subLangs = opts.SubLangs
		}
		args = append(args, "--write-subs", "--sub-langs", subLangs)
	}
	if opts.WriteThumbnail {
		args = append(args, "--write-thumbnail")
	}

	if err := os.MkdirAll(opts.OutputDir, 0o755); err != nil {
		return "", err
	}

	cmd := exec.CommandContext(ctx, opts.YtdlpBin, args...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return "", err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return "", err
	}

	if err := cmd.Start(); err != nil {
		return "", err
	}

	var finalFilepath string
	var destinations []string
	var lastStderr string
	part := 0
	totalParts := 1
	lastDownloaded := float64(0)
	var playlistItem, playlistTotal int

	go func() {
		scanner := bufio.NewScanner(stderr)
		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, "ERROR:") {
				lastStderr = line
			}
		}
	}()

	scanner := bufio.NewScanner(stdout)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		if m := playlistItemRegex.FindStringSubmatch(line); m != nil {
			playlistItem, _ = strconv.Atoi(m[1])
			playlistTotal, _ = strconv.Atoi(m[2])
		}

		if strings.HasPrefix(line, progressTag) {
			body := strings.TrimPrefix(line, progressTag)
			parts := strings.Split(body, "|")
			if len(parts) >= 5 {
				dlVal := parseProgressFloat(parts[0])
				downloaded := float64(0)
				if dlVal != nil {
					downloaded = *dlVal
				}
				if downloaded < lastDownloaded {
					part++
				}
				lastDownloaded = downloaded

				total := parseProgressFloat(parts[1])
				if total == nil {
					total = parseProgressFloat(parts[2])
				}
				speed := float64(0)
				if s := parseProgressFloat(parts[3]); s != nil {
					speed = *s
				}
				eta := float64(0)
				if e := parseProgressFloat(parts[4]); e != nil {
					eta = *e
				}

				if handlers.OnProgress != nil {
					handlers.OnProgress(BakeProgress{
						DownloadedBytes: downloaded,
						TotalBytes:      total,
						Speed:           speed,
						ETA:             eta,
						Part:            part,
						TotalParts:      totalParts,
						PlaylistItem:    playlistItem,
						PlaylistTotal:   playlistTotal,
					})
				}
			}
		} else if aria := ParseAriaProgressLine(line); aria != nil {
			if aria.DownloadedBytes < lastDownloaded && aria.DownloadedBytes > 0 {
				part++
			}
			lastDownloaded = aria.DownloadedBytes

			etaVal := float64(0)
			if aria.ETA != nil {
				etaVal = *aria.ETA
			}
			if handlers.OnProgress != nil {
				handlers.OnProgress(BakeProgress{
					DownloadedBytes: aria.DownloadedBytes,
					TotalBytes:      aria.TotalBytes,
					Speed:           aria.Speed,
					ETA:             etaVal,
					Part:            part,
					TotalParts:      totalParts,
					PlaylistItem:    playlistItem,
					PlaylistTotal:   playlistTotal,
					Connections:     aria.Connections,
					Seeders:         aria.Seeders,
				})
			}
		}

		if m := formatCountRegex.FindStringSubmatch(line); m != nil {
			totalParts = len(strings.Split(m[1], "+"))
		} else if strings.Contains(line, "[Merger]") || strings.Contains(line, "[ExtractAudio]") {
			if m := mergerTargetRegex.FindStringSubmatch(line); m != nil {
				destinations = append(destinations, m[1])
			} else if m := extractAudioTargetRegex.FindStringSubmatch(line); m != nil {
				destinations = append(destinations, m[1])
			}
			if handlers.OnProcessing != nil {
				handlers.OnProcessing()
			}
		} else if strings.HasPrefix(line, "[download] Destination: ") {
			dest := strings.TrimPrefix(line, "[download] Destination: ")
			destinations = append(destinations, dest)
		} else if filepath.IsAbs(line) {
			finalFilepath = line
		}
	}

	err = cmd.Wait()
	if ctx.Err() != nil {
		for _, d := range destinations {
			_ = os.Remove(d)
			_ = os.Remove(d + ".part")
			_ = os.Remove(d + ".ytdl")
		}
		return "", fmt.Errorf("download cancelled")
	}

	if err != nil {
		clean := CleanErrorOutput(lastStderr)
		if clean != "" {
			return "", fmt.Errorf("%s", clean)
		}
		return "", fmt.Errorf("download failed: %w", err)
	}

	if finalFilepath != "" {
		if opts.IsPlaylist {
			return filepath.Dir(finalFilepath), nil
		}
		return finalFilepath, nil
	}

	return opts.OutputDir, nil
}
